### PR TITLE
Modify to a contiguous memory map file using multiple files

### PR DESCRIPTION
**Pull request by MichaelKim20**
A memory map file consists of several files.
Previously, there was a gap between the files. We got rid of it and connected it as one.
And when we save a block or read a block, we can serialize it from the inside.
We can expect better performance because we can reduce memory replication.
MMFile

### CODE REVIEW

File: source/agora/common/BlockStorage.d [hunk 1]
```diff
@@ -382,6 +397,50 @@ public class BlockStorage
...
+    ***************************************************************************/
+
+    public bool readBlockMerkleRoot (ref Block block, Hash merkle_root) @safe
+    {
+        ubyte[HASH_SIZE] hash_bytes = merkle_root[0..HASH_SIZE];
+
+        auto finds
+            = this.merkle_root_idx[].find!((a, b) => a.hash == b)(hash_bytes);
+
+        if (!finds.empty)
```
Comment (AndrejMitrovic) [inline source/agora/common/BlockStorage.d]: It's better to return early. So:

```D
if (finds.empty)
    return false;
```

And then the rest of the code can be unindented by one level. It makes it easier to reason about code this way.

Comment (MichaelKim20) [reply source/agora/common/BlockStorage.d]: Yes, I'll modify it.

File: source/agora/common/BlockStorage.d [hunk 2]
```diff
@@ -781,5 +781,5 @@ version(none) unittest
         assert(random_block.header.merkle_root == merkle_roots[idx]);
     }
 
-    other.release();
+    other.unMap();
```
Comment (Geod24) [inline source/agora/common/BlockStorage.d]: Sorry, I noticed that after. I thought `release` was unused and so removed it. But it is a better name, all things considered, so the integration test PR fixes this.

**Comment by Geod24** [comment]
You should rebase on v0.x.x

**Comment by MichaelKim20** [comment]
A memory map file consists of several files.
Previously, there was a gap between the files. We got rid of it and connected it as one.
And when we save a block or read a block, we can serialize it from the inside.
We can expect better performance because we can reduce memory replication.

**Comment by MichaelKim20** [comment]
Std.mmfile was modified to obtain the address of the memory map file.
However, it was difficult to secure adjacent memory continuously.

So whenever a new memory map is created. 
It was possible to change the memory address of the base point.

However, I thought it was not appropriate to have to modify std.mmfile.
This was resolved by the concept of position of the file.

Then, when accessing the portion at the edge of the file in the memory map, it was done one by one.

I think I've met all the requirements now.

File: source/agora/common/BlockStorage.d [hunk 3]
```diff
@@ -33,17 +33,8 @@ import std.path;
...
 
 
-/// The size of header
-private immutable size_t MFILE_HEAD_SIZE = size_t.sizeof;
-
-/// Amount of reserved size
-private immutable size_t MFILE_RESERVE_SIZE = 64 * 1024;
-
-/// The maximum number of block in one file
-private immutable ulong MFILE_MAX_BLOCK = 100;
```
Comment (Geod24) [inline source/agora/common/BlockStorage.d]: In this commit (the second) those constants are removed, not renamed, so this commit does not compile.

File: source/agora/common/BlockStorage.d [hunk 4]
```diff
@@ -113,73 +113,37 @@ public class BlockStorage
...
-                    0,
+                    MapSize,
                     null
                 );
-            this.data_size = this.readDataLength();
+
+            this.file_base = MapSize * this.file_index;
         }
-        else
+        catch (Exception ex)
```
Comment (Geod24) [inline source/agora/common/BlockStorage.d:147]: So the failure does not get reported to the caller. Does that mean we are okay with this component failing ?

File: source/agora/common/BlockStorage.d [hunk 5]
```diff
@@ -149,19 +149,24 @@ public class BlockStorage
...
     public void release () @trusted
     {
-        if (this.file is null)
-            return;
         import core.memory : GC;
 
-        destroy(this.file);
-        GC.free(&this.file);
-        this.file = null;
+        try
```
Comment (Geod24) [inline source/agora/common/BlockStorage.d]: Did you see `release` throw ? Because it should never happen.

File: source/agora/common/BlockStorage.d [hunk 6]
```diff
@@ -244,6 +244,7 @@ public class BlockStorage
 
         return true;
     }
+
```
Comment (Geod24) [inline source/agora/common/BlockStorage.d:257]: Unrelated change (commit 926779701294e16eba7c744344ad30f976d02033)

File: source/agora/common/BlockStorage.d [hunk 7]
```diff
@@ -359,12 +360,33 @@ public class BlockStorage
...
+            (to / MapSize == this.file_index))
+        {
+            ubyte[] data;
+            try
+            {
+                const size_t x0 = from - this.file_base;
+                const size_t x1 = to - this.file_base;
+                data = cast(ubyte[])this.file[x0 .. x1];
+            }
+            catch (Exception ex)
```
Comment (Geod24) [inline source/agora/common/BlockStorage.d:417]: Flat-out ignoring an exception does not seem like the best way to go

File: source/agora/common/BlockStorage.d [hunk 8]
```diff
@@ -377,13 +399,76 @@ public class BlockStorage
...
+                foreach (idx, e; data)
+                    this.file[x0+idx] = e;
+            }
+            else
+            {
+                foreach (idx, e; data)
+                    this.writeByte(pos + idx, e);
+            }
+        }
+        catch (Exception ex)
```
Comment (Geod24) [inline source/agora/common/BlockStorage.d:467]: Same here, why are we ignoring Exception ?

**Review by Geod24 — CHANGES_REQUESTED** [review]
Good commit separation. The review was much simpler. Thanks for that 👍 
My recommendation: The commit title should indicate what & why something is being done. A commit message like "Modify function read, write" is not informative. "Remove ability to read unmapped file" gives much more info, for example (that's just an example). However, things like "Remove unused variable xxx" or "Fix typo in documentation", for example, are fine because it is self evident from the diff.
Last thing, there have been a lot of empty `try` / `catch` added. They seem to have been added to make the function `nothrow`, however that means that errors are now ignored, and we cannot guarantee the integrity of the data on disk, which is pretty bad.

**Comment by Geod24** [comment]
Can you submit the first 4 commits in an independent PR ? They are good to go.

File: source/agora/common/BlockStorage.d (hunk 8 above)
Comment (MichaelKim20) [reply source/agora/common/BlockStorage.d:467]: fixed

File: source/agora/common/BlockStorage.d (hunk 4 above)
Comment (MichaelKim20) [reply source/agora/common/BlockStorage.d:147]: fixed

File: source/agora/common/BlockStorage.d (hunk 3 above)
Comment (MichaelKim20) [reply source/agora/common/BlockStorage.d]: fixed

File: source/agora/common/BlockStorage.d (hunk 8 above)
Comment (MichaelKim20) [reply source/agora/common/BlockStorage.d:467]: fixed

File: source/agora/common/BlockStorage.d (hunk 7 above)
Comment (MichaelKim20) [reply source/agora/common/BlockStorage.d:417]: fixed

File: source/agora/common/BlockStorage.d (hunk 5 above)
Comment (MichaelKim20) [reply source/agora/common/BlockStorage.d]: fixed

File: source/agora/common/BlockStorage.d (hunk 3 above)
Comment (MichaelKim20) [reply source/agora/common/BlockStorage.d]: fixed

File: source/agora/common/BlockStorage.d (hunk 4 above)
Comment (MichaelKim20) [reply source/agora/common/BlockStorage.d:147]: fixed

**Comment by Geod24** [comment]
@MukeunKim :
Please do not ignore `Exception`.
The pattern:
```d
try { /* ... */ }
catch (Exception e) {}
```
Is very bad because if an exception is triggered, we won't know.
I understand that is done because there are some `nothrow` in place.
In this case, for the time being, I recommend to use:
```d
void func () nothrow
{
    scope (failure) assert(0);
    // ...
}
```

This will make the function `nothrow`

**Comment by Geod24** [comment]
Ah, it looks like the last commit should be merged with the first one. But I recommend to just `assert` for the time being.

**Comment by Geod24** [comment]
Ready for a new review

**Comment by AndrejMitrovic** [comment]
What about the test-case including multiple transactions in the block?

**Comment by MukeunKim** [comment]
@AndrejMitrovic

**Comment by MichaelKim20** [comment]
Added test casings for multi-transaction to block.
I have also verified normal operation for large blocks.
I'm trying to add @AndrejMitrovic  to coworker for the test code, but it doesn't work.

**Comment by AndrejMitrovic** [comment]
> I'm trying to add @AndrejMitrovic to coworker for the test code, but it doesn't work.

It doesn't matter, I just want us to have a test.

File: source/agora/common/BlockStorage.d [hunk 9]
```diff
@@ -301,8 +280,17 @@ public class BlockStorage
         if (finds.empty)
             return false;
 
-        this.readBlockAtPosition(block, finds.front.position);
-        return true;
+        bool success = true;
```
Comment (AndrejMitrovic) [inline source/agora/common/BlockStorage.d]: In general, it is much better to assign failure by default, not success.

However there is no need for this boolean. This can be rewritten to:

```D
try
{
    this.readBlockAtPosition(block, finds.front.position);
    return true;
}
catch (Exception ex)
{
    this.writeLog("BlockStorage.readBlock: ", ex);
    return false;
}
```

File: source/agora/common/BlockStorage.d [hunk 10]
```diff
@@ -328,26 +316,33 @@ public class BlockStorage
...
+        try
+        {
+            this.readBlockAtPosition(block, finds.front.position);
```
Comment (AndrejMitrovic) [inline source/agora/common/BlockStorage.d:320]: Same comment here as above.

File: source/agora/common/BlockStorage.d [hunk 11]
```diff
@@ -441,30 +457,79 @@ public class BlockStorage
...
-            return 0;
+        ubyte data;
+        try
+        {
+            if (this.map(pos / MapSize))
+                data = this.file[pos - this.file_base];
+        }
+        catch (Exception ex)
+        {
+            writeLog("BlockStorage.readBytes: ", ex);
```
Comment (AndrejMitrovic) [inline source/agora/common/BlockStorage.d:497]: So in case of a failure, it will return a ubyte with the value 0 and the error will be ignored?

File: source/agora/common/BlockStorage.d [hunk 12]
```diff
@@ -441,30 +457,79 @@ public class BlockStorage
...
+    private void writeByte (size_t pos, ubyte data) nothrow @trusted
+    {
+        try
+        {
+            if (this.map(pos / MapSize))
+                this.file[pos - this.file_base] = data;
+        }
+        catch (Exception ex)
+        {
+            this.writeLog("BlockStorage.writeByte: ", ex);
```
Comment (AndrejMitrovic) [inline source/agora/common/BlockStorage.d:525]: Why is it ok to fail to write and then ignore the error? If the node can't write to disk, wouldn't it be better to abort the whole process and shut down the node? The data that was written will be invalid when its being read later, because some of the data is missing (the checksum will fail, once it's implemented).

File: source/agora/common/BlockStorage.d [hunk 13]
```diff
@@ -441,30 +457,79 @@ public class BlockStorage
...
+            }
+            else
+            {
+                foreach (idx, e; data)
+                    this.writeByte(pos + idx, e);
+            }
+        }
+        catch (Exception ex)
+        {
+            this.writeLog("BlockStorage.write: ", ex);
```
Comment (AndrejMitrovic) [inline source/agora/common/BlockStorage.d:469]: Also same comment about ignoring write errors.

**Comment by MichaelKim20** [comment]
Review Please!

**Comment by AndrejMitrovic** [comment]
It looks great now. How about adding `nothrow` everywhere where there a function returns a status code?

**Comment by AndrejMitrovic** [comment]
@MukeunKim if you get too many errors trying to make them `nothrow`, or if it's too much work, then leave it the way it is. We can add `nothrow` later.

**Review by AndrejMitrovic — APPROVED** [review]

**Comment by MukeunKim** [comment]
Ready for a new review

**Comment by MichaelKim20** [comment]
@Geod24 Ready for a new review ^^

File: source/agora/common/BlockStorage.d [hunk 14]
```diff
@@ -119,79 +116,44 @@ public class BlockStorage
...
-        If a size change occurs, open it again.
 
         Params:
             findex = the index of the file.
-            resize = The size of the file to change. If 0 does not change it.
 
     ***************************************************************************/
 
-    private void map (size_t findex, size_t resize = 0) @trusted
+    private bool map (size_t findex) @trusted nothrow
```
Comment (Geod24) [inline source/agora/common/BlockStorage.d:125]: Missing `Returns:`

**Review by Geod24 — APPROVED** [review]
LGTM
